File: leetcode/h100/group_1/code03/longest_seq.go
package main

import "fmt"

func main() {
	nums := []int{100, 4, 200, 1, 3, 2}
	ans := longestConsecutive(nums)
	fmt.Println(ans)
}

/*给定一个未排序的整数数组 nums ，找出数字连续的最长序列（不要求序列元素在原数组中连续）的长度。

请你设计并实现时间复杂度为 O(n) 的算法解决此问题。

 */
func longestConsecutive(nums []int) int {
	ans := 0
	set := map[int]bool{}
	for _, num := range nums {
		set[num] = true
	}
	for num := range set {
		if !set[num-1] {
			cur := num
			len := 1
			for set[cur+1] {
				cur++
				len++
			}
			if ans < len {
				ans = len
			}
		}
	}
	return ans
}

File: leetcode/hot100-java/group_1/longest_seq.java
/*
 * 128. 最长连续序列
 *
 * 给定一个未排序的整数数组 nums，找出数字连续的最长序列长度；序列中的元素
 * 不要求在原数组中相邻。算法需要达到线性时间复杂度。
 *
 * 算法实现说明：
 * 1. 先把所有数字放入哈希集合，既能去重，也能在常数平均时间内判断某个数字是否存在。
 * 2. 只有当 num - 1 不在集合中时，才把 num 当作一段连续序列的起点，避免从序列中间重复计数。
 * 3. 从起点不断查询下一个整数并累计长度，最后用所有序列长度更新 longest。
 *
 * 时间复杂度：O(n)，每个不同元素至多被连续扫描一次；空间复杂度：O(n)。
 */
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> values = new HashSet<>();
        for (int num : nums) {
            values.add(num);
        }

        int longest = 0;
        for (int num : values) {
            if (!values.contains(num - 1)) {
                int current = num;
                int length = 1;
                while (values.contains(current + 1)) {
                    current++;
                    length++;
                }
                longest = Math.max(longest, length);
            }
        }
        return longest;
    }
}
